### PR TITLE
SqlServer: Translate TimeSpan members

### PR DESCRIPTION
**Pull request by aleksandar-manukov**
This PR is related to the issue #19632 .

The following translations are enabled:

| .NET                                 | Sql                                                         |
| --------------------------- | ------------------------------------------ |
| timeSpan.Hours               | `DATEPART(hour, @timeSpan)`              |
| timeSpan.Minutes            | `DATEPART(minute, @timeSpan)`          |
| timeSpan.Seconds           | `DATEPART(second, @timeSpan)`          |
| timeSpan.Milliseconds     | `DATEPART(millisecond, @timeSpan)`   |

### CODE REVIEW

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs [hunk 1]
```diff
@@ -7570,6 +7570,46 @@ public virtual Task Group_by_nullable_property_and_project_the_grouping_key_HasV
...
         }
 
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task TimeSpan_Hours(bool async)
+        {
+            return AssertQueryScalar(
+                async,
+                ss => ss.Set<Mission>()
+                    .Select(g => g.Duration.Hours));
```
Comment (roji) [inline test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: Shouldn't we prefer to test with Where instead of Select? Otherwise unimplementing providers will falsely pass because of client evaluation in top-level projection...

Comment (smitpatel) [reply test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: We should have tests for both Where & Select.
Where validates server eval.
Select validates that return typeMapping is set correctly to read back data. (also this should make sure that results are non-empty.

File: src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            _sqlExpressionFactory = sqlExpressionFactory;
+        }
+
+        public virtual SqlExpression Translate(SqlExpression instance, MemberInfo member, Type returnType)
+        {
+            Check.NotNull(member, nameof(member));
+            Check.NotNull(returnType, nameof(returnType));
+
+            SqlExpression translatedSqlExpression = null;
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs]: Remove variable declaration & direct return.

File: src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            if (member.DeclaringType == typeof(TimeSpan) && _datePartMappings.TryGetValue(member.Name, out string value))
+            {
+                translatedSqlExpression = _sqlExpressionFactory.Function("DATEPART", new SqlExpression[2]
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs]: simpler `new [] { }`

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs (hunk 1 above)
Comment (roji) [reply test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: @smitpatel just for my understanding, given that the data being read back here is an int and not a TimeSpan (and the Where test would also test non-empty results), would Select tests actually add value?

@bricelam made the point that writing Where tests could be tricky because of varying precision for TimeSpan across databases, at least for milliseconds etc...

Comment (smitpatel) [reply test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: In such case Select can be skipped. Just make sure to add when types are differing, like if above members return int/long depending on member in same translator, it can easily slip in a bug if not careful.
Bool/DateTime series/any weird type which would cause issue while reading.

We could write a fuzzy Where tests like we do for double subtraction. (range check rather than exact)

File: src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs (hunk 3 above)
Comment (aleksandar-manukov) [reply src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs]: This is ready.

File: src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs (hunk 2 above)
Comment (aleksandar-manukov) [reply src/EFCore.SqlServer/Query/Internal/SqlServerTimeSpanMemberTranslator.cs]: This is ready.

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs (hunk 1 above)
Comment (aleksandar-manukov) [reply test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: I added the where tests for TimeSpan members and also fix old tests related to Mission entity which have been failing because of the added property Duration

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs [hunk 4]
```diff
@@ -7570,6 +7570,86 @@ public virtual Task Group_by_nullable_property_and_project_the_grouping_key_HasV
                     .Select(g => g.Key.HasValue));
         }
 
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task TimeSpan_Hours(bool async)
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: Probably also need to skip some of these tests in Sqlite provider. InMemory should work.

Comment (aleksandar-manukov) [reply test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: This is ready. Tests for TimeSpan members are now skipped for Sqlite

File: test/EFCore.Sqlite.FunctionalTests/Query/GearsOfWarQuerySqliteTest.cs [hunk 5]
```diff
@@ -148,6 +148,54 @@ public override async Task Byte_array_filter_by_SequenceEqual(bool async)
 WHERE ""s"".""Banner5"" = @__byteArrayParam_0");
         }
 
+        [ConditionalTheory(Skip = "PR #19774")]
+        public override Task TimeSpan_Hours(bool async)
```
Comment (roji) [inline test/EFCore.Sqlite.FunctionalTests/Query/GearsOfWarQuerySqliteTest.cs]: nit: can use expression-bodied method for all these (even on same line if it fits)

Comment (aleksandar-manukov) [reply test/EFCore.Sqlite.FunctionalTests/Query/GearsOfWarQuerySqliteTest.cs]: This is done 😉

**Comment by bricelam** [comment]
Why didn't Cosmos fail? Is the whole class skipped there?

**Comment by smitpatel** [comment]
> Why didn't Cosmos fail? Is the whole class skipped there?

Yes. Cosmos does not implement gearsofwar tests.
